### ISSUE TITLE
Discussion regarding `kwargs`

### ISSUE THREAD

**Issue by santisoler** [simpeg/simpeg#1420]
I think it would be nice to continue the discussion regarding the usage of `**kwargs` in class constructors and their role in inheritance.

In past meetings I've expressed my desire to avoid the usage of `**kwargs` arguments in class constructors unless we find a very good reason to do so. I think there are a few different reasons to do so:

1. Allowing `kwargs` on constructors increases complexity of class abstractions. Passing arguments as kwargs through the inheritance tree can have hard-to-track behaviors that depend mostly on the order in which those arguments are being passed. I've experienced these behaviors while refactoring the base classes in #1241.
1. The `kwargs` could possible lead to misbehavior. 
    - If the passed arguments are ignored, the class will behave in an unexpected way without complain, leading to a silent misbehavior (as seen in https://github.com/simpeg/user-tutorials/issues/12).
    - If the class assigns every `kwarg` as attribute, then the class is a container of everything, breaking the sense of abstraction (as the `BaseObjectiveFunction` was doing before https://github.com/simpeg/simpeg/pull/1241/files#diff-57dc55ee40104c9badf8c244fea415e8d563ccf942370e706505f4a155dc2cc3L36)  
1. Interfacing functions with `kwargs` is hard. Users cannot rely on autocompletion or type checking tools when passing arguments through `kwargs`. Moreover, as a developer, the possible arguments that could be passed to `kwargs` are usually overlooked, generating incomplete docstrings. Moreover, asking required arguments to be passed as `kwargs` can create objects that will fail down the line. A good example of this are some simulations like [gravity.Simulation3DIntegral](http://docs.simpeg.xyz/content/api/generated/SimPEG.potential_fields.gravity.Simulation3DIntegral.html#SimPEG.potential_fields.gravity.Simulation3DIntegral), that can be created without passing a `survey`, but they will ultimately fail to run `dpred()`.
1. Keeping `kwargs` might lead to more maintenance work while deprecating arguments. If a function accepts `kwargs`, we need to write an error raising if a deprecated argument is being passed through `kwargs`. When to remove that error is not a trivial question (since we don't want to suddenly allow users to pass legacy arguments).

I think most of these issues can be solved by removing `**kwargs` from the signatures of methods that don't need them, and explicitly ask for the required arguments either as positional or optional (with a default value).

**Comment by santisoler** [comment]
@jcapriot made a point during today's meeting regarding the second issue I mention. That is to add a `super().__init__(**kwargs)` in the topmost parent class. This way, instantiation would fail if there's any residual kwarg that reaches that class.

For example:

```python
class ParentOne:
    def __init__(self, x, **kwargs):
        super().__init__(**kwargs)
        self.x = x


class ChildOne(ParentOne):
    def __init__(self, x, y, **kwargs):
        super().__init__(x, **kwargs)
        self.y = y
```

If we try to instantiate the `ChildOne` class with an extra argument, we get:

```python
ChildOne(1, 2, argument=3)
```
```
TypeError                                 Traceback (most recent call last)
Cell In[2], line 1
----> 1 ChildOne(1, 2, argument=3)

File ~/tmp/sandbox/testing.py:9, in ChildOne.__init__(self, x, y, **kwargs)
      8 def __init__(self, x, y, **kwargs):
----> 9     super().__init__(x, **kwargs)
     10     self.y = y

File ~/tmp/sandbox/testing.py:3, in ParentOne.__init__(self, x, **kwargs)
      2 def __init__(self, x, **kwargs):
----> 3     super().__init__(**kwargs)
      4     self.x = x

TypeError: object.__init__() takes exactly one argument (the instance to initialize)
```

On the other hand, if we avoid `kwargs`, we can define these two classes:

```python
class ParentTwo:
    def __init__(self, x):
        self.x = x


class ChildTwo(ParentTwo):
    def __init__(self, x, y):
        super().__init__(x)
        self.y = y
```

And if we instantiate the `ChildTwo` with an extra non-valid argument, we get:
```python
ChildTwo(1, 2, argument=3)
```
```
TypeError                                 Traceback (most recent call last)
Cell In[3], line 1
----> 1 ChildTwo(1, 2, argument=3)

TypeError: ChildTwo.__init__() got an unexpected keyword argument 'argument'
```

I think the second approach is much more verbose and easier to debug than the first one, where the error is coming from the parent class without further details on which is the offending argument.

**Comment by jcapriot** [comment]
In general, it is tough to remove `**kwargs` from `__init__` scripts everywhere in SimPEG because we make use of multiple inheritance in many class. (See IP simulations, EquivalentSource simulations, etc.).

To me it does make sense that the classes we intend users to work with (anything that is not a `Base` Class) should have them be explicit though.

**Comment by santisoler** [comment]
You're right, it would be very complicated to remove `**kwargs` everywhere. I don't think we should tackle this all at once. But I think we should still discuss this topic for having a decision towards new code, refactors, etc.

I agree that removing `**kwargs` from public classes is a great way to start.

**Comment by lheagy** [comment]
There are a number of places where I have noticed people getting tripped up with `**kwargs` (most recently in some of the receiver classes). I am supportive of taking a first step and removing from public classes (and doing so in stages).

**Comment by jcapriot** [comment]
Just be aware that you need to keep in mind that we don't just use linear inheritance as in your example. Many of the simulation classes have diamond inheritance i.e. consider this typical structure:

```Python
class BasePDE():
    def __init__(mesh, **kwargs):
        self.mesh = mesh
        # Maybe we use the BasePDE as a mixin for some other class, so still would need to call super...
        super().__init__(**kwargs)

class BaseElectricalPDESimulation(BasePDE):
    def __init__(conductivity, **kwargs):
        super().__init__(**kwargs)
        self.conductivity = conductivity

class BaseMagneticPDE(BasePDE):
    def __init__(permeability, **kwargs):
        super().__init__(**kwargs)
        self.permeability = permeability

class ElectromagneticPDE(BaseElectricalPDE, BaseMagneticPDE):
    def __init__(mesh, conductivity, permeability):
        super().__init__(mesh=mesh, conductivity=conductivity, permeability=permeability)
```

The challenge would be to do this without using `**kwargs`, in way that only causes `BasePDE.__init__` to be called once, while making sure that `BaseMagnetic` and `BaseElectrical` still initialize their parent when used individually.

**Comment by jcapriot** [comment]
I also feel it's then cumbersome as a developer to add a feature to a base class, then have to go find every class that inherits from it and modify their call signatures.

**Comment by jcapriot** [comment]
As far as documenting the `kwargs` inherited from parents, I know matplotlib has automatic generators for their inherited portions, https://github.com/matplotlib/matplotlib/blob/main/lib/matplotlib/_docstring.py#L78, that we could take some inspiration from.

**Comment by jcapriot** [comment]
For each base class that has `object` as a parent, we could add a call that checks if the call to `super()` will call object, and return a better error message.
```Python
def _check_base_super(self, this_class, **kwargs):
    mro = type(self).__mro__
    super_class = mro[mro.index(this_class) + 1]
    if super_class is object:
        if len(kwargs):
            error_string = ""
            for key in kwargs:
                error_string += f"{key} is not a support keyword argument of class {type(self)}. "
            error_string = error_string[:-1]
            raise TypeError(error_string)
```

which could be used as:
```Python
class BasePDE():
    def __init__(self, mesh, **kwargs):
        _check_base_super(self, __class__, **kwargs)
        super().__init__(**kwargs)
        print('BasePDE')
        self.mesh = mesh
```

**Comment by jcapriot** [comment]
> For each base class that has `object` as a parent, we could add a call that checks if the call to `super()` will call object, and return a better error message.
> 
> ```python
> def _check_base_super(self, this_class, **kwargs):
>     mro = type(self).__mro__
>     super_class = mro[mro.index(this_class) + 1]
>     if super_class is object:
>         if len(kwargs):
>             for key in kwargs:
>                 raise TypeError(f"{type(sim).__name__} got an unexpected keyword argument '{key}'.")
> ```
> 
> which could be used as:
> 
> ```python
> class BasePDE():
>     def __init__(self, mesh, **kwargs):
>         _check_base_super(self, __class__, **kwargs)
>         super().__init__(**kwargs)
>         print('BasePDE.__init__')
>         self.mesh = mesh
> ```

In fact this is something we should probably do now to save a bit of headache for unsupported keyword arguments until we resolve all of this discussion.

**Comment by santisoler** [comment]
> @santisoler Just be aware that you need to keep in mind that we don't just use linear inheritance as in your example. Many of the simulation classes have diamond inheritance i.e. consider this typical structure:
> 
> ```python
> class BasePDE():
>     def __init__(self, mesh, **kwargs):
>         self.mesh = mesh
>         # Maybe we use the BasePDE as a mixin for some other class, so still would need to call super...
>         super().__init__(**kwargs)
> 
> class BaseElectricalPDE(BasePDE):
>     def __init__(self, conductivity, **kwargs):
>         super().__init__(**kwargs)
>         self.conductivity = conductivity
> 
> class BaseMagneticPDE(BasePDE):
>     def __init__(self, permeability, **kwargs):
>         super().__init__(**kwargs)
>         self.permeability = permeability
> 
> class ElectromagneticPDE(BaseElectricalPDE, BaseMagneticPDE):
>     def __init__(self, mesh, conductivity, permeability):
>         super().__init__(mesh=mesh, conductivity=conductivity, permeability=permeability)
> ```
> 
> The challenge would be to do this without using `**kwargs`, in way that only causes `BasePDE.__init__` to be called once, while making sure that `BaseMagnetic` and `BaseElectrical` still initialize their parent when used individually.

I agree that `**kwargs` is one of the best options for solving diamond inheritance. I'm not against `kwargs` in general, and I totally support them if there's a very good reason for them.

I'm just raising the concern that even in these simple lines there's a level of complexity that most people using Python aren't too familiar with (mro, multiple inheritance, mixins, etc). My goal when raising this discussion it to reduce the level of complexity throughout the codebase where possible. But obviously, some implementations are best solved with more complex code, and that's fine.
